substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -715,40 +779,96 @@ pub mod pallet {
-      let next_validators =
+      let session = Self::session(NetworkId::Serai).unwrap();
+      let next_next_validators =
         Participants::<T>::get(NetworkId::Serai).expect("no Serai participants after new_session");
       Babe::<T>::enact_epoch_change(
         WeakBoundedVec::force_from(
-          validators.iter().copied().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
+          next_validators.iter().copied().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
           None,
         ),
         WeakBoundedVec::force_from(
-          next_validators.into_iter().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
+          next_next_validators.into_iter().map(|(id, w)| (BabeAuthorityId::from(id), w)).collect(),
           None,
         ),
-        Some(session),
+        Some(session.0),
       );
       Grandpa::<T>::new_session(
         true,
-        session,
-        validators.into_iter().map(|(id, w)| (GrandpaAuthorityId::from(id), w)).collect(),
+        session.0,
+        next_validators.into_iter().map(|(id, w)| (GrandpaAuthorityId::from(id), w)).collect(),
       );
     }
 
+    // Since babe & grandpa authorities are the same each session, we can check whichever.
+    // Babe also provides convenient function to check for authorities so we just go with that.
+    fn can_slash_serai_validator(validator: &Public) -> bool {
+      Babe::<T>::is_member(&BabeAuthorityId::from(*validator)) ||
+        Self::previous_participants(NetworkId::Serai)
+          .into_iter()
+          .any(|(id, _)| id.into_inner() == *validator)
+    }
+
+    fn slash_serai_validator(validator: &Public) {
+      let network = NetworkId::Serai;
+      let account = *validator;
+      let mut allocation = Self::allocation((network, account)).unwrap();
+
+      // reduce the current allocation to 0.
+      Self::set_allocation(network, account, Amount(0));
+
+      // prevent any future deallocation.
+      let mut to_unlock_on = Self::session(network).unwrap();
+      // TODO: we don't know in which block the it might be deallocating, so we go through all
+      // but this is ugly.
+      for i in 0 .. 4 {
+        to_unlock_on.0 += i;
+        let pending = Self::take_deallocatable_amount(network, to_unlock_on, account);
+        if pending.is_some() {
+          allocation = allocation.add(pending.unwrap());
+          PendingDeallocations::<T>::set((network, to_unlock_on, account), None);
+        }
+      }
+
+      // remove from the current & future sets.
+      // TODO: how to remove from the current?
+      let mut participants =
+        Participants::<T>::get(network).expect("no Serai participants on slash");
+      let index = participants.iter().position(|(id, _)| *id == account);
+      if let Some(i) = index {
+        participants.remove(i);
+        Participants::<T>::set(network, Some(participants));
+        InSet::<T>::remove(Self::in_set_key(network, account));
+      }
```

Comment:
Also, there's probably some unhandled side effects for the signals pallet here.

Reply:
`pallet_babe` exposes `DisabledValidators` type that we implement [here](https://github.com/serai-dex/serai/blob/c511a54d18453814719155b30777c99bcf355f7f/substrate/validator-sets/pallet/src/lib.rs#L1004). The thing is yes, all it does is to "panic" if one of those disabled ones produces a block. So surely somewhere they prevent picking these validators from producing a block? Otherwise whole network would halt when one of them does. Because I don't see that code in the runtime-babe and so I didn't use that feature atm.